File: ShipModels.cs
/// ============================================================
/// 拉格朗日AI — C# 类库
/// 编译：csc /target:library /out:LagrangeLib.dll ShipModels.cs
/// 用途：.NET客户端舰船数据模型
/// ============================================================

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LagrangeAI.Models
{
    /// <summary>舰船类型枚举</summary>
    public enum ShipType
    {
        [JsonPropertyName("battleship")] Battleship,
        [JsonPropertyName("battlecruiser")] Battlecruiser,
        [JsonPropertyName("aircraftcarrier")] AircraftCarrier,
        [JsonPropertyName("support")] Support,
        [JsonPropertyName("cruiser")] Cruiser,
        [JsonPropertyName("destroyer")] Destroyer,
        [JsonPropertyName("frigate")] Frigate,
        [JsonPropertyName("fighter")] Fighter,
        [JsonPropertyName("corvette")] Corvette
    }

    /// <summary>评级等级</summary>
    public enum Rating { S = 10, A = 7, B = 4, C = 2, D = 0 }

    /// <summary>舰船数据模型</summary>
    public class ShipData
    {
        [JsonPropertyName("id")] public string Id { get; set; } = "";
        [JsonPropertyName("name")] public string Name { get; set; } = "";
        [JsonPropertyName("variant")] public string Variant { get; set; } = "";
        [JsonPropertyName("type")] public string Type { get; set; } = "";
        [JsonPropertyName("hp")] public long Hp { get; set; }
        [JsonPropertyName("physicalArmor")] public int PhysicalArmor { get; set; }
        [JsonPropertyName("energyArmor")] public int EnergyArmor { get; set; }
        [JsonPropertyName("commandValue")] public int CommandValue { get; set; }
        [JsonPropertyName("ratings")] public Dictionary<string, string>? Ratings { get; set; }

        public string FullName => string.IsNullOrEmpty(Variant) ? Name : $"{Name}{Variant}";
        public bool IsSuperCapital => Type is "battleship" or "battlecruiser" or "aircraftcarrier" or "support";

        /// <summary>计算综合战斗力评分</summary>
        public double CalculateCombatScore()
        {
            double hpScore = Hp / 10000.0 * 3;
            double armorScore = PhysicalArmor / 20.0 * 2;
            double shieldScore = EnergyArmor / 10.0;

            double ratingScore = 0;
            if (Ratings != null)
            {
                var scoreMap = new Dictionary<string, int> { {"S",10},{"A",7},{"B",4},{"C",2},{"D",0} };
                foreach (var r in Ratings.Values)
                    ratingScore += scoreMap.GetValueOrDefault(r, 0);
            }

            double efficiency = Hp / Math.Max(CommandValue, 1.0) / 1000.0 * 5;
            return hpScore + armorScore + shieldScore + ratingScore + efficiency;
        }

        public string GetTypeName() => Type switch
        {
            "battleship" => "战列舰",
            "battlecruiser" => "战列巡洋舰",
            "aircraftcarrier" => "航空母舰",
            "support" => "支援舰",
            "cruiser" => "巡洋舰",
            "destroyer" => "驱逐舰",
            "frigate" => "护卫舰",
            "fighter" => "战机",
            "corvette" => "护航艇",
            _ => Type
        };

        public override string ToString() =>
            $"[{GetTypeName()}] {FullName} | HP:{Hp:N0} | CV:{CommandValue} | Score:{CalculateCombatScore():F1}";
    }

    /// <summary>舰队配置</summary>
    public class FleetConfig
    {
        public List<ShipData> MainFleet { get; set; } = new();
        public List<ShipData> Reinforcement { get; set; } = new();
        public string? FlagshipId { get; set; }
        public int MaxCV { get; set; } = 500;

        public int TotalCV => MainFleet.Sum(s => s.CommandValue) + Reinforcement.Sum(s => s.CommandValue);
        public bool IsValid => TotalCV <= MaxCV;
        public long TotalHP => MainFleet.Sum(s => s.Hp) + Reinforcement.Sum(s => s.Hp);
    }

    /// <summary>API客户端（简化版）</summary>
    public class LagrangeApiClient
    {
        private readonly HttpClient _http;
        private readonly string _baseUrl;
        private string? _token;

        public LagrangeApiClient(string baseUrl = "http://127.0.0.1:3000")
        {
            _baseUrl = baseUrl;
            _http = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        }

        public async Task<bool> HealthCheckAsync()
        {
            var resp = await _http.GetAsync($"{_baseUrl}/health");
            return resp.IsSuccessStatusCode;
        }

        public async Task<int> GetShipCountAsync()
        {
            var resp = await _http.GetAsync($"{_baseUrl}/api/ships");
            var json = await resp.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.GetProperty("count").GetInt32();
        }
    }
}

File: ShipModels.swift
// ============================================================
// 拉格朗日AI — Swift 数据模型
// 用途：iOS/macOS Swift 客户端舰船数据结构
// ============================================================

import Foundation

/// 舰船评级
enum ShipRating: String, Codable, CaseIterable {
    case S, A, B, C, D
    
    var score: Int {
        switch self {
        case .S: return 10
        case .A: return 7
        case .B: return 4
        case .C: return 2
        case .D: return 0
        }
    }
}

/// 舰船类型
enum ShipType: String, Codable, CaseIterable {
    case battleship, battlecruiser, aircraftcarrier, support
    case cruiser, destroyer, frigate, fighter, corvette
    
    var chineseName: String {
        switch self {
        case .battleship: return "战列舰"
        case .battlecruiser: return "战列巡洋舰"
        case .aircraftcarrier: return "航空母舰"
        case .support: return "支援舰"
        case .cruiser: return "巡洋舰"
        case .destroyer: return "驱逐舰"
        case .frigate: return "护卫舰"
        case .fighter: return "战机"
        case .corvette: return "护航艇"
        }
    }
    
    var isSuperCapital: Bool {
        switch self {
        case .battleship, .battlecruiser, .aircraftcarrier, .support:
            return true
        default:
            return false
        }
    }
}

/// 舰船速度
struct ShipSpeed: Codable {
    let cruise: Int
    let warp: Int
}

/// 舰船数据模型
struct LagrangeShip: Codable, Identifiable {
    let id: String
    let name: String
    let variant: String
    let type: ShipType
    let hp: Int
    let physicalArmor: Int
    let energyArmor: Int
    let commandValue: Int
    let ratings: [String: ShipRating]
    let speed: ShipSpeed?
    
    var fullName: String {
        variant.isEmpty ? name : "\(name)\(variant)"
    }
    
    /// 综合战斗力评分
    var combatScore: Double {
        let hpScore = Double(hp) / 10000.0 * 3
        let armorScore = Double(physicalArmor) / 20.0 * 2
        let shieldScore = Double(energyArmor) / 10.0
        let ratingScore = ratings.values.reduce(0) { $0 + $1.score }
        let efficiency = Double(hp) / max(Double(commandValue), 1) / 1000.0 * 5
        return hpScore + armorScore + shieldScore + Double(ratingScore) + efficiency
    }
    
    // MARK: - API 客户端
    
    /// 从API获取所有舰船
    static func fetchAll(from baseURL: String = "http://127.0.0.1:3000") async throws -> [LagrangeShip] {
        guard let url = URL(string: "\(baseURL)/api/ships") else {
            throw URLError(.badURL)
        }
        
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        
        struct ShipsResponse: Codable {
            let ships: [LagrangeShip]
            let count: Int
        }
        
        let decoder = JSONDecoder()
        let result = try decoder.decode(ShipsResponse.self, from: data)
        return result.ships
    }
    
    /// 健康检查
    static func healthCheck(baseURL: String = "http://127.0.0.1:3000") async -> Bool {
        guard let url = URL(string: "\(baseURL)/health") else { return false }
        do {
            let (_, response) = try await URLSession.shared.data(from: url)
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }
}

// MARK: - 舰队配置

struct LagrangeFleet {
    var name: String
    var ships: [LagrangeShip]
    var flagship: LagrangeShip?
    let maxCommandValue: Int = 500
    
    var totalCV: Int { ships.reduce(0) { $0 + $1.commandValue } }
    var totalHP: Int { ships.reduce(0) { $0 + $1.hp } }
    var isValid: Bool { totalCV <= maxCommandValue }
    
    /// 按评分排序
    var topShips: [LagrangeShip] {
        ships.sorted { $0.combatScore > $1.combatScore }
    }
}
